pkg/server/test/expand.go

```diff
@@ -810,53 +533,244 @@ func TestExpandQuery(t *testing.T, datastore storage.OpenFGADatastore) {
+								Value: &openfgav1.UsersetTree_Leaf_Users{
+									Users: &openfgav1.UsersetTree_Users{
+										Users: []string{"document:2#editor"},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "self_defined_userset_not_returned",
+			model: testutils.MustTransformDSLToProtoWithID(`
+			model
+				schema 1.1
+			type user
+			type group
+				relations
+					define viewer: [user]
+			`),
+			tuples: []*openfgav1.TupleKey{},
+			request: &openfgav1.ExpandRequest{
+				TupleKey: tuple.NewExpandRequestTupleKey("group:1", "viewer"),
+			},
+			expected: &openfgav1.ExpandResponse{
+				Tree: &openfgav1.UsersetTree{
+					Root: &openfgav1.UsersetTree_Node{
+						Name: "group:1#viewer",
+						Value: &openfgav1.UsersetTree_Node_Leaf{
+							Leaf: &openfgav1.UsersetTree_Leaf{
+								Value: &openfgav1.UsersetTree_Leaf_Users{
+									Users: &openfgav1.UsersetTree_Users{
+										// group:1#viewer isn't included because it's implicit
+										Users: []string{},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "self_defined_userset_not_returned_even_if_tuple_written",
+			model: testutils.MustTransformDSLToProtoWithID(`
+			model
+				schema 1.1
+			type user
+			type group
+				relations
+					define viewer: [user]
+			`),
+			tuples: []*openfgav1.TupleKey{
+				tuple.NewTupleKey("group:1", "viewer", "group:1#viewer"), // invalid, so should be skipped over
+			},
+			request: &openfgav1.ExpandRequest{
+				TupleKey: tuple.NewExpandRequestTupleKey("group:1", "viewer"),
+			},
+			expected: &openfgav1.ExpandResponse{
+				Tree: &openfgav1.UsersetTree{
+					Root: &openfgav1.UsersetTree_Node{
+						Name: "group:1#viewer",
+						Value: &openfgav1.UsersetTree_Node_Leaf{
+							Leaf: &openfgav1.UsersetTree_Leaf{
+								Value: &openfgav1.UsersetTree_Leaf_Users{
+									Users: &openfgav1.UsersetTree_Users{},
+								},
+							},
+						},
 					},
-					{
-						Type: "document",
-						Relations: map[string]*openfgav1.Userset{
-							"parent": typesystem.This(),
-							"editor": typesystem.This(),
+				},
+			},
+		},
+		{
+			name: "self_defined_userset_returned_if_tuple_written",
+			model: testutils.MustTransformDSLToProtoWithID(`
+			model
+				schema 1.1
+			type user
+			type group
+				relations
+					define viewer: [user, group#viewer]
+			`),
+			tuples: []*openfgav1.TupleKey{
+				tuple.NewTupleKey("group:1", "viewer", "group:1#viewer"),
+			},
+			request: &openfgav1.ExpandRequest{
+				TupleKey: tuple.NewExpandRequestTupleKey("group:1", "viewer"),
+			},
+			expected: &openfgav1.ExpandResponse{
+				Tree: &openfgav1.UsersetTree{
+					Root: &openfgav1.UsersetTree_Node{
+						Name: "group:1#viewer",
+						Value: &openfgav1.UsersetTree_Node_Leaf{
+							Leaf: &openfgav1.UsersetTree_Leaf{
+								Value: &openfgav1.UsersetTree_Leaf_Users{
+									Users: &openfgav1.UsersetTree_Users{
+										Users: []string{"group:1#viewer"},
```

Comment:
I think this is the case that we will want to address in https://github.com/openfga/openfga/pull/1521, right @rhamzeh ?

Reply:
Yes... and no :D

This should return empty, if we're assuming it is always truthy and blocking writing that tuple, so if the tuple can no longer be written, expand will not return it but it will be assumed true.

But I don't think we should make this change without blocking writing those tuples

Reply:
Ah, yes. In https://github.com/openfga/openfga/pull/1521 I explicitly forbid writing that tuple.

Reply:
I think for that case, if it's an existing tuple I'm fine with returning it.

The only reason I'm saying that is that I would prefer expand has no exceptions we need to maintain. If it's in the DB, it's fair game. Does that make sense?